Doesn't work in local (where-bound) function?

The following code fails to typecheck:

```
{-# LANGUAGE GADTs, ScopedTypeVariables, DataKinds, KindSignatures #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE Rank2Types #-}

{-# OPTIONS_GHC -fplugin GHC.TypeLits.KnownNat.Solver #-}

import Data.Singletons.Prelude.Num
import Data.Singletons
import GHC.TypeLits

data Foo :: Nat -> * where
  FZ  :: Foo 0
  FS :: Foo n -> Foo (n + 1)

sPred :: (KnownNat n) => Sing (n+1) -> Sing n
sPred _ = sing

foo :: forall a n. (KnownNat n) => a -> (forall m. KnownNat m => Sing m -> a) -> Foo n -> a
foo x0 f = go (sing :: Sing n)
  where
    go :: forall k. (KnownNat k) => Sing k -> Foo k -> a
    go s FZ = x0
    go s (FS _) = f (sPred s)
```

The error message is as follows (on GHC 8.0.2):

```
KnownNatSolverBug.hs:25:19: error:
    * Could not deduce (KnownNat n1) arising from a use of `f'
      from the context: KnownNat n
        bound by the type signature for:
                   foo :: KnownNat n =>
                          a -> (forall (m :: Nat). KnownNat m => Sing m -> a) ->
        at KnownNatSolverBug.hs:20:1-91
      or from: KnownNat k
        bound by the type signature for:
                   go :: KnownNat k => Sing k -> Foo k -> a
        at KnownNatSolverBug.hs:23:5-56
      or from: k ~ (n1 + 1)
        bound by a pattern with constructor:
                   FS :: forall (n :: Nat). Foo n -> Foo (n + 1),
                 in an equation for `go'
        at KnownNatSolverBug.hs:25:11-14
      Possible fix:
        add (KnownNat n1) to the context of the data constructor `FS'
    * In the expression: f (sPred s)
      In an equation for `go': go s (FS _) = f (sPred s)
      In an equation for `foo':
          foo x0 f
            = go (sing :: Sing n)
            where
                go :: forall k. (KnownNat k) => Sing k -> Foo k -> a
                go s FZ = x0
                go s (FS _) = f (sPred s)
```

To highlight, the relevant parts of the context are:

```
    * Could not deduce (KnownNat n1) arising from a use of `f'
      from: KnownNat k
      or from: k ~ (n1 + 1)
```

But isn't this exactly the situation that the plugin was designed to handle?

Actually `foo` above can be simplified further to just:

```
foo :: forall a n. (KnownNat n) => a -> (forall m. KnownNat m => Sing m -> a) -> Sing n -> Foo n -> a
foo x0 f s FZ = x0
foo x0 f s (FS _) = f (sPred s)
```

```
KnownNatSolverBug.hs:22:21: error:
    * Could not deduce (KnownNat n1) arising from a use of `f'
      from the context: KnownNat n
      or from: n ~ (n1 + 1)
```

The following example, even simpler, cuts right to the crux of the issue:

```
{-# LANGUAGE DataKinds, TypeFamilies, TypeOperators #-}
{-# LANGUAGE Rank2Types, FlexibleContexts #-}

{-# OPTIONS_GHC -fplugin GHC.TypeLits.KnownNat.Solver #-}

import Data.Singletons.Prelude.Num
import Data.Singletons
import GHC.TypeLits

bar :: (KnownNat n, n ~ (m+1)) => (forall k. KnownNat k => Sing k -> a) -> Sing m -> a
bar f s = f s
```

```
bug.hs:11:11: error:
    • Could not deduce (KnownNat m) arising from a use of ‘f’
      from the context: (KnownNat n, n ~ (m + 1))
        bound by the type signature for:
                   bar :: (KnownNat n, n ~ (m + 1)) =>
                          (forall (k :: Nat). KnownNat k => Sing k -> a) -> Sing m -> a
        at bug.hs:10:1-86

```

Just ran into a similar issue:

```haskell
{-# LANGUAGE AllowAmbiguousTypes, DataKinds, TypeFamilies, TypeOperators #-}
{-# OPTIONS_GHC -fplugin GHC.TypeLits.KnownNat.Solver #-}
{-# OPTIONS_GHC -fplugin GHC.TypeLits.Normalise #-}
{-# OPTIONS_GHC -dcore-lint #-}

module EqTest where

import Numeric.Natural
import GHC.TypeNats

foo :: (KnownNat n, KnownNat m, k ~ (n + m)) => proxy k -> Natural
foo = natVal
```

This produces the following error:

```
src/EqTest.hs:12:7: error:
    • Could not deduce (KnownNat k) arising from a use of ‘natVal’
      from the context: (KnownNat n, KnownNat m, k ~ (n + m))
        bound by the type signature for:
                   foo :: forall (n :: Nat) (m :: Nat) (k :: Nat) (proxy :: Nat -> *).
                          (KnownNat n, KnownNat m, k ~ (n + m)) =>
                          proxy k -> Natural
        at src/EqTest.hs:11:1-66
      Possible fix:
        add (KnownNat k) to the context of
          the type signature for:
            foo :: forall (n :: Nat) (m :: Nat) (k :: Nat) (proxy :: Nat -> *).
                   (KnownNat n, KnownNat m, k ~ (n + m)) =>
                   proxy k -> Natural
    • In the expression: natVal
      In an equation for ‘foo’: foo = natVal
   |
12 | foo = natVal
   |       ^^^^^^
```
I'm on GHC 8.6.5, and using ``ghc-typelits-knownnat`` and ``ghc-typelits-natnormalise`` version 0.7.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,8 @@
 
 ## 0.8.1
 * Fix [#53](https://github.com/clash-lang/ghc-typelits-knownnat/issues/53) The plugin sometimes doesn't look through type aliases
+* Fix [#13](https://github.com/clash-lang/ghc-typelits-knownnat/issues/13) Type equality constraints aren't used by solver
+* Fix [#42](https://github.com/clash-lang/ghc-typelits-knownnat/issues/42) Intermediate type variable stops derivation of `KnownNat` constraint
 
 ## 0.8.0 *September 8th 2025*
 * Support for GHC 9.14.1.

--- a/src/GHC/TypeLits/KnownNat/Solver.hs
+++ b/src/GHC/TypeLits/KnownNat/Solver.hs
@@ -105,6 +105,10 @@ where
 -- base
 import Control.Arrow
   ( (&&&), first )
+import Data.Foldable
+  ( asum )
+import Data.List.NonEmpty as NE
+  ( filter )
 import Data.Maybe
   ( catMaybes, fromMaybe, mapMaybe )
 
@@ -117,7 +121,7 @@ import Control.Monad.Trans.Writer.Strict
 import GHC.TypeLits.Normalise.SOP
   ( SOP (..), Product (..), Symbol (..) )
 import GHC.TypeLits.Normalise.Unify
-  ( CType (..),normaliseNat, reifySOP, CoreSOP )
+  ( CType (..),normaliseNat, reifySOP )
 
 -- ghc-tcplugin-api
 import GHC.TcPlugin.API
@@ -339,16 +343,31 @@ constraintToEvTerm defs givensTyConSubst givens (ct,cls,op,orig) = do
                  found@Just {} -> return found
                  -- 3.b If not, we check if the outer type-level operation
                  -- has a corresponding KnownNat<N> instance.
-                 _ -> go (op,Nothing)
+                 _ -> go [] (op,Nothing)
     return ((first (,ct)) <$> evM)
   where
     -- Determine whether the outer type-level operation has a corresponding
     -- KnownNat<N> instance, where /N/ corresponds to the arity of the
     -- type-level operation
-    go :: (Type, Maybe Coercion) -> TcPluginM Solve (Maybe (EvTerm,[Ct]))
-    go (coreView -> Just tyN, coM) = go (tyN, coM)
-    go (go_other -> Just ev, _) = return (Just (ev,[]))
-    go (ty@(TyConApp tc args0), sM)
+    go :: [Coercion] -> (Type, Maybe Coercion) -> TcPluginM Solve (Maybe (EvTerm,[Ct]))
+    -- Look through type aliases
+    go deps (coreView -> Just tyN, coM) = go deps (tyN, coM)
+    -- Look through rewrites
+    go deps0 (ty, coM)
+      | Just tcapps <- splitTyConApp_upTo givensTyConSubst ty
+      -- We are only interested in the splitTyConApp_upTo result that used a
+      -- rewrite
+      , withDeps@(_:_) <- NE.filter (\(_,_,deps) -> not (null deps)) tcapps
+      = do results <- traverse (\(tc, args, deps1) -> go (deps0 <> deps1)
+                                                         (TyConApp tc args, coM))
+                               withDeps
+           return (asum results)
+    -- See whether there is a given that matches it (after having looked through
+    -- type aliases and rewrites)
+    go deps (go_other deps -> Just ev, _) = return (Just (ev,[]))
+    -- And if there isn't, see whether we can construct it using a KnownNat<N>
+    -- instance
+    go deps (ty@(TyConApp tc args0), sM)
       | let tcNm = tyConName tc
       , Just m <- nameModule_maybe tcNm
       = do
@@ -400,8 +419,6 @@ constraintToEvTerm defs givensTyConSubst givens (ct,cls,op,orig) = do
                         . splitFunTys          -- ([KnownNat x, KnowNat y], DKnownNat2 "+" x y)
                         . (`piResultTys` args0N) -- (KnowNat x, KnownNat y) => DKnownNat2 "+" x y
                         $ idType df_id         -- forall a b . (KnownNat a, KnownNat b) => DKnownNat2 "+" a b
-                deps :: [Coercion]
-                deps = [] -- XXX TODO: not declaring dependency on outer Givens
             (evs,new) <- unzip <$> mapM (go_arg . irrelevantMult) df_args
             if className cls == className (knownBool defs)
                -- Create evidence using the original, flattened, argument of
@@ -440,17 +457,17 @@ constraintToEvTerm defs givensTyConSubst givens (ct,cls,op,orig) = do
                -- 'KnownNat fsk'; the one GHC originally asked us to solve.
                then return ((,concat new) <$> makeOpDictByFiat df cls args1N args0N (unOrig orig) deps evs)
                else return ((,concat new) <$> makeOpDict df cls args1N args0N (unOrig orig) deps evs (fmap (ty,) sM))
-          _ -> return ((,[]) <$> go_other ty)
+          _ -> return ((,[]) <$> go_other deps ty)
 
-    go ((LitTy (NumTyLit i)), _)
+    go deps ((LitTy (NumTyLit i)), _)
       -- Let GHC solve simple Literal constraints
       | LitTy _ <- op
       = return Nothing
       -- This plugin only solves Literal KnownNat's that needed to be normalised
       -- first
       | otherwise
-      = (fmap (,[])) <$> makeLitDict cls op [] i -- XXX: ok to pass empty dependent coercions?
-    go _ = return Nothing
+      = (fmap (,[])) <$> makeLitDict cls op deps i
+    go _ _ = return Nothing
 
     -- Get EvTerm arguments for type-level operations. If they do not exist
     -- as [G]iven constraints, then generate new [W]anted constraints
@@ -463,13 +480,13 @@ constraintToEvTerm defs givensTyConSubst givens (ct,cls,op,orig) = do
 
     -- Fall through case: look up the normalised [W]anted constraint in the list
     -- of [G]iven constraints.
-    go_other :: Type -> Maybe EvTerm
-    go_other ty =
+    go_other :: [Coercion] -> Type -> Maybe EvTerm
+    go_other deps ty =
       let knClsTc = classTyCon cls
           kn      = mkTyConApp knClsTc [ty]
           cast    = if CType ty == CType op
                        then Just . EvExpr
-                       else makeKnCoercion cls ty op [] -- XXX: ok to pass empty dependent coercions?
+                       else makeKnCoercion cls ty op deps
       in  cast =<< lookup (CType kn) givens
 
     -- Find a known constraint for a wanted, so that (modulo normalization)
@@ -507,19 +524,16 @@ constraintToEvTerm defs givensTyConSubst givens (ct,cls,op,orig) = do
           -- with the original Nat operation
           subWant  = mkTyConApp typeNatSubTyCon . (:[want])
           -- exploded :: [()]
-          exploded = map (discardCo . runWriter . normaliseNat givensTyConSubst . subWant . fst &&& id)
+          exploded = map (fst . runWriter . normaliseNat givensTyConSubst . subWant . fst &&& id)
                          knownsX
-          -- XXX TODO: discarding coercions produced by 'normaliseNat'
-          discardCo :: ((CoreSOP, [Coercion]), [(Type, Type)]) -> CoreSOP
-          discardCo ((a, _co), _) = a
           -- interesting cases for us are those where
           -- wanted and given only differ by a constant
-          examineDiff (S [P [I n]]) entire = Just (entire,I n)
-          examineDiff (S [P [V v]]) entire = Just (entire,V v)
+          examineDiff ((S [P [I n]]),deps) entire = Just (entire,I n,deps)
+          examineDiff ((S [P [V v]]),deps) entire = Just (entire,V v,deps)
           examineDiff _ _ = Nothing
           interesting = mapMaybe (uncurry examineDiff) exploded
       -- convert the first suitable evidence
-      (((h,sM),corr):_) <- pure interesting
+      (((h,sM),corr,deps):_) <- pure interesting
       x <- case corr of
                 I 0 -> pure (fromMaybe (h,Nothing) sM)
                 I i | i < 0
@@ -573,7 +587,7 @@ constraintToEvTerm defs givensTyConSubst givens (ct,cls,op,orig) = do
                             ( mkTyConApp typeNatSubTyCon [h,lC]
                             , Nothing
                             )
-      MaybeT (go x)
+      MaybeT (go deps x)
 
 makeWantedEv
   :: Ct

--- a/tests/Main.hs
+++ b/tests/Main.hs
@@ -201,6 +201,12 @@ repeatV = undefined
 test29 :: KnownNat x => Vec (NatTimes2 x) Bool
 test29 = repeatV False
 
+test30 :: forall a b . (b ~ (2^a)) => SNat a -> SNat (Log b)
+test30 SNat = SNat @(Log b)
+
+test31 :: (KnownNat n, KnownNat m, k ~ (n + m)) => proxy n -> proxy m -> proxy k -> Natural
+test31 _ _ = natVal
+
 tests :: TestTree
 tests = testGroup "ghc-typelits-natnormalise"
   [ testGroup "Basic functionality"
@@ -277,6 +283,12 @@ tests = testGroup "ghc-typelits-natnormalise"
     , testCase "KnownNat (a + b), KnownNat b => KnownNat a; @(a+b) ~ 8, b ~ 6" $
       show (test21 (Proxy @8) (Proxy @6)) @?=
       "2"
+    , testCase "b ~ 2 ^ a, KnownNat a => KnownNat (Log b)" $
+      show (test30 (SNat @8)) @?=
+      "8"
+    , testCase "k ~ m + n, KnownNat m, KnownNat n => KnownNat k" $
+      show (test31 (Proxy @2) (Proxy @6) Proxy) @?=
+      "8"
     ],
     testGroup "Normalisation"
     [ testCase "KnownNat (m-n+n) ~ KnownNat m" $

